File: ts/src/lens.ts
/**
 * Lens (SPEC §9): the canonical, compact text a model reads instead of raw JSON.
 * Deterministic: two conforming implementations produce byte-identical output.
 */
import { quote } from "./canonical.js";
import type { Effect, Money, More, ParamSchema, Proposal, Reply } from "./types.js";

const BARE = /^[A-Za-z0-9_@./+\-:() '!?&%$#*=<>~^]+$/;
const NUMERIC = /^-?(0|[1-9]\d*)(\.\d+)?([eE][+-]?\d+)?$/;
const RESERVED = new Set(["-", "true", "false", "null"]);
const pad = (n: number) => "  ".repeat(n);

const isScalar = (v: unknown): v is string | number | boolean | null =>
  v === null || typeof v === "string" || typeof v === "number" || typeof v === "boolean";
const isObject = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);

export function scalar(v: unknown): string {
  if (v === null || v === undefined) return "-";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return Number.isFinite(v) ? String(v) : "-";
  const s = String(v);
  if (BARE.test(s) && s[0] !== " " && s[s.length - 1] !== " " && !RESERVED.has(s) && !NUMERIC.test(s)) return s;
  return quote(s);
}

function isTable(arr: unknown[]): arr is Record<string, unknown>[] {
  if (!arr.every(isObject)) return false;
  const first = Object.keys(arr[0]);
  if (first.length === 0) return false;
  return arr.every((o) => {
    const ks = Object.keys(o);
    return ks.length === first.length && ks.every((k, i) => k === first[i]) && ks.every((k) => isScalar(o[k]));
  });
}

function entries(obj: Record<string, unknown>, n: number): string[] {
  const out: string[] = [];
  for (const [k, v] of Object.entries(obj)) if (v !== undefined) out.push(...entry(k, v, n));
  return out;
}

function entry(key: string, v: unknown, n: number): string[] {
  const k = pad(n) + scalar(key);
  if (isScalar(v)) return [`${k}: ${scalar(v)}`];
  if (Array.isArray(v)) {
    if (v.length === 0) return [`${k}: []`];
    if (v.every(isScalar)) return [`${k}: [${v.map(scalar).join(", ")}]`];
    if (isTable(v)) {
      const cols = Object.keys(v[0]);
      return [`${k}[${v.length}]{${cols.map(scalar).join(",")}}:`, ...v.map((row) => pad(n + 1) + cols.map((c) => scalar(row[c])).join(","))];
    }
    return [`${k}[${v.length}]:`, ...v.flatMap((item) => listItem(item, n + 1))];
  }
  if (isObject(v)) {
    const body = entries(v, n + 1);
    return body.length ? [`${k}:`, ...body] : [`${k}: {}`];
  }
  return [`${k}: -`];
}

function listItem(item: unknown, n: number): string[] {
  const dash = pad(n) + "- ";
  if (isScalar(item)) return [dash + scalar(item)];
  if (Array.isArray(item)) return [dash + (item.every(isScalar) ? `[${item.map(scalar).join(", ")}]` : JSON.stringify(item))];
  if (isObject(item)) {
    const body = entries(item, n + 1);
    if (!body.length) return [dash + "{}"];
    return [dash + body[0].slice(pad(n + 1).length), ...body.slice(1)];
  }
  return [dash + "-"];
}

/** Render any JSON value in lean notation (SPEC §9.1). */
export function lean(v: unknown): string {
  if (isScalar(v)) return scalar(v);
  if (Array.isArray(v)) return entry("items", v, 0).join("\n");
  if (isObject(v)) {
    const lines = entries(v, 0);
    return lines.length ? lines.join("\n") : "{}";
  }
  return "-";
}

// ---- formatting helpers (SPEC §9.2) ----

export function fmtTime(unix: number): string {
  const iso = new Date(unix * 1000).toISOString(); // YYYY-MM-DDTHH:MM:SS.sssZ
  const secs = iso.slice(17, 19);
  return iso.slice(0, 16) + (secs === "00" ? "" : ":" + secs) + "Z";
}

export function fmtDuration(s: number): string {
  if (s !== 0 && s % 86400 === 0) return `${s / 86400}d`;
  if (s !== 0 && s % 3600 === 0) return `${s / 3600}h`;
  if (s !== 0 && s % 60 === 0) return `${s / 60}m`;
  return `${s}s`;
}

const ZERO_DECIMAL = new Set(["JPY", "KRW", "VND", "CLP", "ISK", "UGX", "XAF", "XOF"]);

export function fmtMoney(m: Money | null | undefined): string {
  if (!m) return "free";
  if (ZERO_DECIMAL.has(m.currency)) return `${m.amount} ${m.currency}`;
  const neg = m.amount < 0 ? "-" : "";
  const a = Math.abs(m.amount);
  return `${neg}${Math.floor(a / 100)}.${String(a % 100).padStart(2, "0")} ${m.currency}`;
}

const SYM: Record<string, string> = { create: "+", update: "~", delete: "-", send: ">", charge: "$", other: "*" };

export function effectLine(e: Effect): string {
  let s = `${SYM[e.op] ?? "*"} ${e.op} ${e.target}${e.field ? "." + e.field : ""}`;
  if (e.from !== undefined || e.to !== undefined) s += `: ${scalar(e.from ?? null)} → ${scalar(e.to ?? null)}`;
  if (e.detail) s += ` — ${e.detail}`;
  return s;
}

export function paramList(params: ParamSchema | undefined): string {
  if (!params) return "()";
  const ty = (t: ParamSchema[string]): string => (typeof t === "string" ? t : Array.isArray(t) ? `[${ty(t[0])}]` : "{" + paramList(t).slice(1, -1) + "}");
  return "(" + Object.entries(params).map(([k, t]) => `${k}: ${ty(t)}`).join(", ") + ")";
}

function moreLines(more: More[] | undefined): string[] {
  return (more ?? []).map((m) => `… ${m.remaining} more at ${m.path} — EXPAND ${m.handle} (~${m.est} tokens)`);
}

const ATTRS: [string, (p: Proposal) => string][] = [
  ["cost", (p) => fmtMoney(p.cost)],
  ["risk", (p) => p.risk],
  ["undo", (p) => (p.undo ? fmtDuration(p.undo.window) : "never")],
  ["expires", (p) => fmtTime(p.expires)],
];

function proposalsLines(ps: Proposal[]): string[] {
  // Attributes identical across all (N ≥ 2) proposals are stated once, in the header.
  const shared = ps.length >= 2 ? ATTRS.filter(([, f]) => ps.every((p) => f(p) === f(ps[0]))) : [];
  const own = ATTRS.filter((a) => !shared.includes(a));
  const out = [`${ps.length} proposal${ps.length === 1 ? "" : "s"}${shared.length ? " — " + shared.map(([k, f]) => `${k}: ${f(ps[0])}`).join(" · ") : ""}:`];
  for (const p of ps) {
    out.push(`[${p.id}] ${p.summary}`, ...p.effects.map((e) => "  " + effectLine(e)));
    if (own.length) out.push("  " + own.map(([k, f]) => `${k}: ${f(p)}`).join(" · "));
    if (p.data !== undefined) out.push(...entry("data", p.data, 1));
  }
  return out;
}

/** Render a reply frame as Lens (SPEC §9.2). */
export function lens(r: Reply): string {
  const out: string[] = [];
  switch (r.kind) {
    case "BRIEF":
      out.push(`# ${r.service.name} (${r.service.id})`);
      if (r.service.summary) out.push(r.service.summary);
      for (const c of r.capabilities) out.push(`${c.kind} ${c.name}${paramList(c.params)}${c.summary ? ` — ${c.summary}` : ""}${c.risk ? ` [risk:${c.risk}]` : ""}`);
      break;
    case "ANSWER":
      out.push(lean(r.data));
      break;
    case "PROPOSALS":
      out.push(...proposalsLines(r.proposals));
      break;
    case "CLARIFY":
      out.push(`? ${r.question}`, ...r.options.map((o, i) => `  ${i + 1}. ${o.label}`));
      break;
    case "RECEIPT": {
      const rc = r.receipt;
      const tag = `(receipt ${rc.id})${r.replay ? " (replay)" : ""}`;
      if (rc.undoes) out.push(`↶ undid ${rc.undoes}: ${rc.summary} ${tag}`);
      else out.push(`✓ ${rc.summary} ${tag} · ${rc.undo ? `undo until ${fmtTime(rc.undo.until)}` : "irreversible"}`);
      // The model already saw the effects in the proposal, unless the service auto-committed.
      if (r.auto) out.push(...rc.effects.map((e) => "  " + effectLine(e)));
      if (rc.result !== undefined) out.push(...entry("result", rc.result, 1));
      break;
    }
    case "ERROR":
      out.push(`✗ ${r.code}: ${r.message}`);
      for (const f of r.fix ?? []) out.push(`  fix: ${f.say}${f.params ? ` → params ${JSON.stringify(f.params)}` : ""}`);
      if (r.need?.length) out.push(`  need: ${JSON.stringify(r.need)}`);
      if (r.consent) out.push(`  consent: principal must approve ${r.consent.hash} (${r.consent.summary})`);
      if (typeof r.retry === "number") out.push(`  retry in: ${fmtDuration(r.retry)}`);
      break;
    case "EVENT":
      out.push(`… ${r.message}${typeof r.progress === "number" ? ` (${Math.round(r.progress * 100)}%)` : ""}`);
      break;
  }
  if ("more" in r) out.push(...moreLines(r.more));
  return out.join("\n");
}

/** Shared token estimate (SPEC §8): ceil(utf8 bytes / 4). */
export const est = (text: string) => Math.ceil(new TextEncoder().encode(text).length / 4);

File: python/src/parley/lens.py
"""Lens: the canonical compact text rendering of values and replies (SPEC §9)."""

from __future__ import annotations

import math
import re
from datetime import datetime, timezone
from typing import Any

from ._json import compact, js_number, quote

_BARE = re.compile(r"[A-Za-z0-9_@./+\-:() '!?&%$#*=<>~^]+")
_JSON_NUMBER = re.compile(r"-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?")
_RESERVED = frozenset({"-", "true", "false", "null"})
_ZERO_DECIMAL = frozenset({"JPY", "KRW", "VND", "CLP", "ISK", "UGX", "XAF", "XOF"})
_EFFECT_SYM = {"create": "+", "update": "~", "delete": "-", "send": ">", "charge": "$", "other": "*"}
_DURATION_UNITS = ((86400, "d"), (3600, "h"), (60, "m"))


def est(text: str) -> int:
    """Shared token estimate: ``ceil(utf8ByteLength / 4)`` (SPEC §8)."""
    return -(-len(text.encode("utf-8")) // 4)


# ---------------------------------------------------------------- lean notation (§9.1)


def _is_scalar(v: Any) -> bool:
    return v is None or isinstance(v, (bool, int, float, str))


def _is_bare(s: str) -> bool:
    return (
        _BARE.fullmatch(s) is not None
        and not s.startswith(" ")
        and not s.endswith(" ")
        and s not in _RESERVED
        and _JSON_NUMBER.fullmatch(s) is None
    )


def scalar(v: Any) -> str:
    if v is None:
        return "-"
    if v is True:
        return "true"
    if v is False:
        return "false"
    if isinstance(v, (int, float)):
        return "-" if isinstance(v, float) and not math.isfinite(v) else js_number(v)
    if isinstance(v, str):
        return v if _is_bare(v) else quote(v)
    return compact(v)


def _ind(n: int) -> str:
    return "  " * n


def _object_lines(obj: dict, n: int) -> list[str]:
    lines: list[str] = []
    for k, v in obj.items():
        lines.extend(_entry_lines(str(k), v, n))
    return lines


def _entry_lines(key: str, v: Any, n: int) -> list[str]:
    key = scalar(key)
    if isinstance(v, dict):
        if not v:
            return [f"{_ind(n)}{key}: {{}}"]
        return [f"{_ind(n)}{key}:"] + _object_lines(v, n + 1)
    if isinstance(v, (list, tuple)):
        return _array_lines(key, list(v), n)
    return [f"{_ind(n)}{key}: {scalar(v)}"]


# _array_lines and _item_lines take an already-rendered key.


def _scalar_list(items: list) -> str:
    return "[" + ", ".join(scalar(x) for x in items) + "]"


def _is_table(items: list) -> bool:
    first = items[0]
    if not isinstance(first, dict) or not first:
        return False
    keys = list(first)
    return all(
        isinstance(x, dict) and list(x) == keys and all(_is_scalar(val) for val in x.values()) for x in items
    )


def _array_lines(key: str, items: list, n: int) -> list[str]:
    if not items:
        return [f"{_ind(n)}{key}: []"]
    if all(_is_scalar(x) for x in items):
        return [f"{_ind(n)}{key}: {_scalar_list(items)}"]
    if _is_table(items):
        keys = list(items[0])
        head = f"{_ind(n)}{key}[{len(items)}]{{{','.join(scalar(k) for k in keys)}}}:"
        rows = [_ind(n + 1) + ",".join(scalar(x[k]) for k in keys) for x in items]
        return [head] + rows
    lines = [f"{_ind(n)}{key}[{len(items)}]:"]
    for el in items:
        lines.extend(_item_lines(el, n + 1))
    return lines


def _item_lines(el: Any, n: int) -> list[str]:
    dash = _ind(n) + "- "
    if isinstance(el, dict):
        if not el:
            return [dash + "{}"]
        sub = _object_lines(el, n + 1)
        sub[0] = dash + sub[0][len(_ind(n + 1)):]
        return sub
    if isinstance(el, (list, tuple)):
        el = list(el)
        return [dash + (_scalar_list(el) if all(_is_scalar(x) for x in el) else compact(el))]
    return [dash + scalar(el)]


def lean(v: Any) -> str:
    """Lean rendering of an arbitrary JSON value."""
    if isinstance(v, dict):
        return "\n".join(_object_lines(v, 0)) or "{}"
    if isinstance(v, (list, tuple)):
        return "\n".join(_array_lines("items", list(v), 0))
    return scalar(v)


# ---------------------------------------------------------------- reply renderings (§9.2)


def fmt_time(t: Any) -> str:
    if type(t) is not int and not (isinstance(t, float) and t.is_integer()):
        return scalar(t)
    d = datetime.fromtimestamp(int(t), tz=timezone.utc)
    s = d.strftime("%Y-%m-%dT%H:%M")
    return s + (f":{d.second:02d}" if d.second else "") + "Z"


def fmt_duration(secs: Any) -> str:
    if type(secs) is not int:
        return scalar(secs)
    for size, unit in _DURATION_UNITS:
        if secs != 0 and secs % size == 0:
            return f"{secs // size}{unit}"
    return f"{secs}s"


def fmt_money(cost: Any) -> str:
    if cost is None:
        return "free"
    amount, currency = cost.get("amount"), cost.get("currency")
    if currency in _ZERO_DECIMAL:
        return f"{amount} {currency}"
    sign = "-" if amount < 0 else ""
    a = abs(amount)
    return f"{sign}{a // 100}.{a % 100:02d} {currency}"


def effect_line(e: dict) -> str:
    op = e.get("op", "other")
    line = f"{_EFFECT_SYM.get(op, '*')} {op} {e.get('target', '')}"
    if e.get("field"):
        line += f".{e['field']}"
    if "from" in e or "to" in e:
        line += f": {scalar(e.get('from'))} → {scalar(e.get('to'))}"
    if e.get("detail"):
        line += f" — {e['detail']}"
    return line


def param_list(params: Any) -> str:
    """``(name: type, …)``; nested schemas render as ``{k: t}``, arrays of them as ``[{…}]``."""
    if not params:
        return "()"

    def ty(t: Any) -> str:
        if isinstance(t, str):
            return t
        if isinstance(t, list):
            return f"[{ty(t[0])}]"
        return "{" + param_list(t)[1:-1] + "}"

    return "(" + ", ".join(f"{k}: {ty(t)}" for k, t in params.items()) + ")"


def _brief(r: dict) -> list[str]:
    svc = r.get("service") or {}
    lines = [f"# {svc.get('name', '')} ({svc.get('id', '')})"]
    if svc.get("summary"):
        lines.append(svc["summary"])
    for c in r.get("capabilities") or []:
        line = f"{c.get('kind', '')} {c.get('name', '')}{param_list(c.get('params'))}"
        if c.get("summary"):
            line += f" — {c['summary']}"
        if c.get("risk"):
            line += f" [risk:{c['risk']}]"
        lines.append(line)
    return lines


def _proposals(r: dict) -> list[str]:
    ps = r.get("proposals") or []
    lines = [f"{len(ps)} proposal{'' if len(ps) == 1 else 's'}:"]
    for p in ps:
        lines.append(f"[{p.get('id', '')}] {p.get('summary', '')}")
        lines.extend("  " + effect_line(e) for e in p.get("effects") or [])
        undo = p.get("undo")
        undo_s = fmt_duration(undo["window"]) if isinstance(undo, dict) and "window" in undo else "never"
        lines.append(
            f"  cost: {fmt_money(p.get('cost'))} · risk: {p.get('risk', '-')} · undo: {undo_s}"
            f" · expires: {fmt_time(p.get('expires'))}"
        )
        if "data" in p:
            lines.extend(_entry_lines("data", p["data"], 1))
    return lines


def _clarify(r: dict) -> list[str]:
    lines = [f"? {r.get('question', '')}"]
    lines.extend(f"  {i}. {o.get('label', '')}" for i, o in enumerate(r.get("options") or [], 1))
    return lines


def _receipt(r: dict) -> list[str]:
    rc = r.get("receipt") or {}
    tail = f"(receipt {rc.get('id', '')})"
    if r.get("replay"):
        tail += " (replay)"
    if rc.get("undoes"):
        lines = [f"↶ undid {rc['undoes']}: {rc.get('summary', '')} {tail}"]
    else:
        lines = [f"✓ {rc.get('summary', '')} {tail}"]
    lines.extend("  " + effect_line(e) for e in rc.get("effects") or [])
    undo = rc.get("undo")
    lines.append(f"  undo: until {fmt_time(undo['until'])}" if isinstance(undo, dict) and "until" in undo else "  undo: never")
    if "result" in rc:
        lines.extend(_entry_lines("result", rc["result"], 1))
    return lines


def _error(r: dict) -> list[str]:
    lines = [f"✗ {r.get('code', '')}: {r.get('message', '')}"]
    for f in r.get("fix") or []:
        line = f"  fix: {f.get('say', '')}"
        if f.get("params") is not None:
            line += f" → params {compact(f['params'])}"
        lines.append(line)
    if r.get("need"):
        lines.append(f"  need: {compact(r['need'])}")
    consent = r.get("consent")
    if consent:
        lines.append(f"  consent: principal must approve {consent.get('hash', '')} ({consent.get('summary', '')})")
    if isinstance(r.get("retry"), (int, float)) and not isinstance(r.get("retry"), bool):
        lines.append(f"  retry in: {fmt_duration(r['retry'])}")
    return lines


def _event(r: dict) -> list[str]:
    line = f"… {r.get('message', '')}"
    if r.get("progress") is not None:
        line += f" ({math.floor(r['progress'] * 100 + 0.5)}%)"  # JS Math.round semantics
    return [line]


def more_line(m: dict) -> str:
    return f"… {m.get('remaining')} more at {m.get('path')} — EXPAND {m.get('handle')} (~{m.get('est')} tokens)"


_RENDERERS = {
    "BRIEF": _brief,
    "PROPOSALS": _proposals,
    "CLARIFY": _clarify,
    "RECEIPT": _receipt,
    "ERROR": _error,
    "EVENT": _event,
    "ANSWER": lambda r: [lean(r.get("data"))],
}


def lens(reply: dict) -> str:
    """Render a reply frame. Ignores any service-supplied ``lens`` field."""
    render = _RENDERERS.get(reply.get("kind", ""))
    lines = render(reply) if render else [lean({k: v for k, v in reply.items() if k not in ("parley", "id", "re")})]
    lines.extend(more_line(m) for m in reply.get("more") or [])
    return "\n".join(lines)

